File: DSA/DSA_450/Cpp/11_heap/013_Check_Binary_Tree_Heap.cpp
/*
Problem: Check if a Binary Tree is a Heap
URL: https://practice.geeksforgeeks.org/problems/is-binary-tree-heap/1

Problem Statement:
Check if a given binary tree satisfies max heap properties: completeness + max heap ordering.

Sample Input/Output:
Input: Tree structure
Output: true/false
*/

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    bool Check_Heap_Recursive(TreeNode* root) {
        /*
        Recursive Approach
        Time Complexity: O(n)
        Space Complexity: O(n)
        */
        int nodeCount = CountNodes(root);
        return IsComplete(root, 0, nodeCount) && IsMaxHeap(root);
    }
    
    bool Check_Heap_Level_Order(TreeNode* root) {
        /*
        Level Order Approach
        Time Complexity: O(n)
        Space Complexity: O(n)
        */
        if (!root) return true;
        
        queue<TreeNode*> q;
        q.push(root);
        bool foundNull = false;
        
        while (!q.empty()) {
            TreeNode* node = q.front();
            q.pop();
            
            if (!node) {
                foundNull = true;
            } else {
                if (foundNull) return false;
                
                if (node->left) {
                    if (node->left->val > node->val) return false;
                    q.push(node->left);
                } else {
                    q.push(nullptr);
                }
                
                if (node->right) {
                    if (node->right->val > node->val) return false;
                    q.push(node->right);
                } else {
                    q.push(nullptr);
                }
            }
        }
        
        return true;
    }
    
private:
    int CountNodes(TreeNode* root) {
        if (!root) return 0;
        return 1 + CountNodes(root->left) + CountNodes(root->right);
    }
    
    bool IsComplete(TreeNode* root, int index, int nodeCount) {
        if (!root) return true;
        if (index >= nodeCount) return false;
        return IsComplete(root->left, 2 * index + 1, nodeCount) &&
               IsComplete(root->right, 2 * index + 2, nodeCount);
    }
    
    bool IsMaxHeap(TreeNode* root) {
        if (!root) return true;
        
        bool leftValid = true;
        bool rightValid = true;
        
        if (root->left) {
            if (root->left->val > root->val) return false;
            leftValid = IsMaxHeap(root->left);
        }
        
        if (root->right) {
            if (root->right->val > root->val) return false;
            rightValid = IsMaxHeap(root->right);
        }
        
        return leftValid && rightValid;
    }
};

void Test_Check_Heap() {
    Solution solution;
    
    TreeNode* root1 = new TreeNode(10);
    root1->left = new TreeNode(9);
    root1->right = new TreeNode(8);
    root1->left->left = new TreeNode(7);
    root1->left->right = new TreeNode(6);
    root1->right->left = new TreeNode(5);
    
    cout << "Test 1 (Valid Heap): " << solution.Check_Heap_Recursive(root1) << endl;
    cout << "Test 1 Level Order: " << solution.Check_Heap_Level_Order(root1) << endl;
    
    TreeNode* root2 = new TreeNode(10);
    root2->left = new TreeNode(15);
    root2->right = new TreeNode(8);
    
    cout << "Test 2 (Invalid Heap): " << solution.Check_Heap_Recursive(root2) << endl;
    cout << "Test 2 Level Order: " << solution.Check_Heap_Level_Order(root2) << endl;
    
    TreeNode* root3 = new TreeNode(10);
    root3->left = new TreeNode(9);
    root3->right = new TreeNode(8);
    root3->left->left = new TreeNode(7);
    
    cout << "Test 3 (Valid Heap): " << solution.Check_Heap_Recursive(root3) << endl;
    cout << "Test 3 Level Order: " << solution.Check_Heap_Level_Order(root3) << endl;
}

int main() {
    Test_Check_Heap();
    return 0;
}

File: DSA/DSA_450/Python/11_heap/013_Check_Binary_Tree_Heap.py
"""
Problem: Check if a Binary Tree is a Heap
URL: https://practice.geeksforgeeks.org/problems/is-binary-tree-heap/1

Problem Statement:
Check if a given binary tree satisfies max heap properties: completeness + max heap ordering.

Sample Input/Output:
Input: Tree structure
Output: true/false
"""

from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def Check_Heap_Recursive(self, root):
        """
        Recursive Approach
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        node_count = self.CountNodes(root)
        return self.IsComplete(root, 0, node_count) and self.IsMaxHeap(root)

    def Check_Heap_Level_Order(self, root):
        """
        Level Order Approach
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        if not root:
            return True

        q = deque([root])
        found_null = False

        while q:
            node = q.popleft()

            if not node:
                found_null = True
            else:
                if found_null:
                    return False

                if node.left:
                    if node.left.val > node.val:
                        return False
                    q.append(node.left)
                else:
                    q.append(None)

                if node.right:
                    if node.right.val > node.val:
                        return False
                    q.append(node.right)
                else:
                    q.append(None)

        return True

    def CountNodes(self, root):
        if not root:
            return 0
        return 1 + self.CountNodes(root.left) + self.CountNodes(root.right)

    def IsComplete(self, root, index, node_count):
        if not root:
            return True
        if index >= node_count:
            return False
        return self.IsComplete(root.left, 2 * index + 1, node_count) and \
               self.IsComplete(root.right, 2 * index + 2, node_count)

    def IsMaxHeap(self, root):
        if not root:
            return True

        left_valid = True
        right_valid = True

        if root.left:
            if root.left.val > root.val:
                return False
            left_valid = self.IsMaxHeap(root.left)

        if root.right:
            if root.right.val > root.val:
                return False
            right_valid = self.IsMaxHeap(root.right)

        return left_valid and right_valid


def Test_Check_Heap():
    solution = Solution()

    root1 = TreeNode(10)
    root1.left = TreeNode(9)
    root1.right = TreeNode(8)
    root1.left.left = TreeNode(7)
    root1.left.right = TreeNode(6)
    root1.right.left = TreeNode(5)

    print("Test 1 (Valid Heap):", solution.Check_Heap_Recursive(root1))
    print("Test 1 Level Order:", solution.Check_Heap_Level_Order(root1))

    root2 = TreeNode(10)
    root2.left = TreeNode(15)
    root2.right = TreeNode(8)

    print("Test 2 (Invalid Heap):", solution.Check_Heap_Recursive(root2))
    print("Test 2 Level Order:", solution.Check_Heap_Level_Order(root2))

    root3 = TreeNode(10)
    root3.left = TreeNode(9)
    root3.right = TreeNode(8)
    root3.left.left = TreeNode(7)

    print("Test 3 (Valid Heap):", solution.Check_Heap_Recursive(root3))
    print("Test 3 Level Order:", solution.Check_Heap_Level_Order(root3))


if __name__ == "__main__":
    Test_Check_Heap()
